Add memcpy32

pdietl opened the pull request.

pdietl commented:
I need @alexanderwachter I need some help testing this.

SebastianBoe commented on include/sys/util.h:
```diff
@@ -363,6 +363,23 @@ static inline uint8_t bin2bcd(uint8_t bin)
+ * This function copies src to dest 32 bits at a time.
+ * If one calls this function without src and dst being aligned
+ * on a 4-byte boundary or if n is not a multiple of 4, it will
+ * do nothing and return NULL.
```
Reading this I thought it was weird there was no error-handling, but I see there are __ASSERT's.

These should be mentioned here.

SebastianBoe commented on lib/os/memcpy32.c:
```diff
@@ -0,0 +1,49 @@
+{
+	__ASSERT(len % 4 == 0, "len must be a multiple of 4!");
+
+	typedef uint32_t __may_alias u32_alias;
```
Scary gcc attribute. But I assume all supported toolchains are in CI.

pdietl replied:
Yes this is scary, but there is no other way to get around strict aliasing that I know of :(

henrikbrixandersen replied:
Should we add `ZRESTRICT` to match `memcpy()` in `lib/libc/minimal/include/string.h`?

pdietl replied:
Done

tejlmand replied:
> Scary gcc attribute. But I assume all supported toolchains are in CI.

actually they are not.

Only the Zephyr SDK and clang is in CI, but none of the other commercial toolchains.

Other toolchains are gcc or clang based, so if this works with those two, I expect other toolchain should work.
@pdietl can you ensure that this will be tested by CI for minimum Zephyr SDK and clang.

SebastianBoe commented on lib/os/memcpy32.c:
```diff
@@ -0,0 +1,49 @@
+
+#define IS_ALIGNED(X) (!((uintptr_t)X & (sizeof(uint32_t) - 1)))
+
+void *memcpy32(void *restrict dst0, const void *restrict src0, size_t len)
```
Strange, I thought restrict was only useful if the same memory address was read multiple times.

But we don't read the same address multiple times in this memcpy.

Did you see any changes to the generated machine code with and without the restrict keyword?

pdietl replied:
Regular `memcpy` uses `restrict` in its parameters, so I did the same here.

pdietl replied:
I removed it anyway

henrikbrixandersen replied:
In Zephyr, you can use `ZRESTRICT`.

SebastianBoe commented on lib/os/CMakeLists.txt:
```diff
@@ -24,6 +24,7 @@ zephyr_sources(
   heap-validate.c
   bitarray.c
   multi_heap.c
+  memcpy32.c
```
Let's have a Kconfig for this to not slow down all builds that don't use it.

pdietl replied:
This doesn't seem necessary. When statically linking, only symbols that are used are kept. If `mcan` is not selected by Kconfig, then `memcpy32` doesn't end up in the zephyr binary.

SebastianBoe replied:
It's so that niche features that the majority don't use don't slow down the build time for everyone.

pdietl replied:
I see

SebastianBoe reviewed (changes requested):
Requesting a Kconfig so we don't build this for the vast majority of builds.

pdietl commented:
@SebastianBoe Should I make `mcan` depend on `memcpy32` or have `mcan` select `memcpy32` ?

henrikbrixandersen commented:
> @SebastianBoe Should I make `mcan` depend on `memcpy32` or have `mcan` select `memcpy32` ?

Use `select`, please.

SebastianBoe commented on drivers/can/Kconfig.mcan:
```diff
@@ -7,6 +7,7 @@ config CAN_MCAN
 	bool
 	select CAN_HAS_CANFD
 	select CAN_HAS_RX_TIMESTAMP
+    select MEMCPY32
```
whitespace is off

alexanderwachter reviewed (approved):
Thanks for this fix @pdietl !
I can go to the rest of m_can in a follow up PR to remove the for loops, commented with do not use memcpy :-)

henrikbrixandersen commented on include/sys/util.h:
```diff
@@ -363,6 +363,26 @@ static inline uint8_t bin2bcd(uint8_t bin)
+ * do nothing and return NULL.
+ *
+ * NOTE: This function uses __ASSERT to enforce the requirements
+ * of each parameter give below.
```
```suggestion
 * of each parameter given below.
```

pdietl committed:
utils: memcpy32

Implement memcpy32, a version of memcpy that only
performs 32 bit word-wise copying.

Signed-off-by: Pete Dietl <petedietl@gmail.com>

pdietl committed:
drivers: mcan

Use memcpy32 for word-wise copying

Fixes #41074

Signed-off-by: Pete Dietl <petedietl@gmail.com>

henrikbrixandersen commented on include/sys/util.h:
```diff
@@ -363,6 +363,26 @@ static inline uint8_t bin2bcd(uint8_t bin)
+ * @return destination pointer or NULL on error
+ *
+ */
+void *memcpy32(void *ZRESTRICT est, const void *ZRESTRICT src, size_t n);
```
```suggestion
void *memcpy32(void *ZRESTRICT dest, const void *ZRESTRICT src, size_t n);
```

pdietl replied:
🤦

henrikbrixandersen reviewed (changes requested):
One minor nit left.

pdietl commented:
@henrikbrixandersen For some reason Twister gets upset about the ZRESTRICT.

henrikbrixandersen commented:
> @henrikbrixandersen For some reason Twister gets upset about the ZRESTRICT.

You need to `#include <toolchain.h>` in `util.h`.

pdietl commented:
@henrikbrixandersen huh that's an interesting error. Plz help :)

pdietl commented:
Hmm the PDF documentation gen seems to work well 
![image](https://user-images.githubusercontent.com/10335942/146600145-fb1bbf25-a5fc-403d-8d76-4f2af4939bec.png)
Bu the HTML fails?

alexanderwachter commented:
@pdietl works on the nucleo_g474re

alexanderwachter reviewed (changes requested):
Does not work on HW

alexanderwachter commented on lib/os/memcpy32.c:
```diff
@@ -0,0 +1,49 @@
+		}
+
+		/* len is 12, 8, 4, or 0 */
+		while (len >= 0) {
```
```suggestion
		while (len > 0) {
```

alexanderwachter replied:
size_t cannot be negative.

tejlmand commented:
> > @SebastianBoe Should I make `mcan` depend on `memcpy32` or have `mcan` select `memcpy32` ?
> 
> Use `select`, please.

Are you sure ?
Are we certain that all toolchains support this:
https://github.com/zephyrproject-rtos/zephyr/blob/310e736211c8570469ad42a9b942c720072982c9/lib/os/memcpy32.c#L18

If not, then we could risk ending in a situation where the `MEMCPY32` must have a `depends on <some toolchain>` which again may risk a:
```
warning: MEMCPY32 (defined at ...) has direct dependencies <SOME_TOOLCHAIN> with value n, but is currently being y-selected by the following symbols:
 - CAN_MCAN (defined at ...), with value y .....

error: Aborting due to Kconfig warnings
```
when using a `select MEMCPY32` here.

As minimum we should ensure Zephyr SDK and `clang` is supporting `__ma_alias`, as the toolchains currently supported er either gcc or clang based.

tejlmand reviewed (changes requested):
When building `tests/drivers/can/canfd/drivers.canfd` for `nucleo_g474re` I get the following errors (warning turned into errors) with twister:
```
/projects/github/ncs/zephyr/drivers/can/can_mcan.c: In function 'can_mcan_init':
/projects/github/ncs/zephyr/drivers/can/can_mcan.c:389:2: error: converting a packed 'struct can_mcan_msg_sram' pointer (alignment 1) to a 'uint32_t' {aka 'unsigned int'} pointer (alignment 4) may result in an unaligned pointer value [-Werror=address-of-packed-member]
  389 |  for (uint32_t *ptr = (uint32_t *)msg_ram;
      |  ^~~
In file included from /projects/github/ncs/zephyr/drivers/can/can_mcan.c:11:
/projects/github/ncs/zephyr/drivers/can/can_mcan.h:148:8: note: defined here
  148 | struct can_mcan_msg_sram {
      |        ^~~~~~~~~~~~~~~~~
/projects/github/ncs/zephyr/drivers/can/can_mcan.c: In function 'can_mcan_get_message':
/projects/github/ncs/zephyr/drivers/can/can_mcan.c:525:15: error: taking address of packed member of 'struct can_mcan_rx_fifo' may result in an unaligned pointer value [-Werror=address-of-packed-member]
  525 |    for (src = fifo[get_idx].data_32,
      |               ^~~~
/projects/github/ncs/zephyr/drivers/can/can_mcan.c: In function 'can_mcan_send':
/projects/github/ncs/zephyr/drivers/can/can_mcan.c:698:11: error: passing argument 1 of 'memcpy32' discards 'volatile' qualifier from pointer target type [-Werror=discarded-qualifiers]
  698 |  memcpy32(&msg_ram->tx_buffer[put_idx].hdr, &tx_hdr,
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /projects/github/ncs/zephyr/drivers/can/can_mcan.c:7:
/projects/github/ncs/zephyr/include/sys/util.h:385:32: note: expected 'void * restrict' but argument is of type 'volatile struct can_mcan_tx_buffer_hdr *'
  385 | void *memcpy32(void *ZRESTRICT dest, const void *ZRESTRICT src, size_t n);
      |                ~~~~~~~~~~~~~~~~^~~~
/projects/github/ncs/zephyr/drivers/can/can_mcan.c:702:9: error: taking address of packed member of 'struct can_mcan_tx_buffer' may result in an unaligned pointer value [-Werror=address-of-packed-member]
  702 |   dst = msg_ram->tx_buffer[put_idx].data_32,
      |         ^~~~~~~
cc1: all warnings being treated as errors

```

SebastianBoe commented:
@tejlmand : If a toolchain does not support this then I assume we will get a reasonable error, so I don't see the need for a depends in the Kconfig.

alexanderwachter commented:
@SebastianBoe @henrikbrixandersen asked for it to speed up compilation time.

tejlmand commented:
> @tejlmand : If a toolchain does not support this then I assume we will get a reasonable error, so I don't see the need for a depends in the Kconfig.

except for twister.
There are plans for getting commercial toolchain support in CI, and twister default runs CMake for dts and Kconfig to determine what should be build, and if a board matches the filter, then the test will build for the board.

So just relying on a sane error message build time could impact twister in bad way.
I'm just trying to minimize the risk of having to go through this at later time.

But first thing I would like to know is whether or not the `__may_alias` is understood by clang, cause if it is, we may not have an immediate problem.

tejlmand commented:
> @tejlmand that should be fixed in #41326

Great, so please rebase this PR, so we can continue.

henrikbrixandersen commented:
See #40644 for inspiration.

henrikbrixandersen commented:
@pdietl Do you plan on picking this up again?

pdietl commented:
Ok what do I need to do to finish this off @henrikbrixandersen

pdietl commented:
Since memcpy32_volatile was added already to `can_mcan.c` do we still want or need this?

SebastianBoe commented:
Does memcpy32_volatile not use compiler-specific features? If so, re-writing this to use volatile would resolve the main problems with this PR I believe.

pdietl commented:
@henrikbrixandersen Do you want me to change this version of memcpy32 to be volatile?
Regardless I need some help fixing the build failures

tejlmand reviewed (changes requested):
Still getting build failures when running `twister` in `tests/drivers/can/canfd/drivers.canfd` for `nucleo_g474re`: 

```
 38%] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/memcpy32.c.obj
/projects/github/ncs/zephyr/lib/os/memcpy32.c: In function 'memcpy32':
/projects/github/ncs/zephyr/lib/os/memcpy32.c:20:14: error: initialization discards 'volatile' qualifier from pointer target type [-Werror=discarded-qualifiers]
   20 |  char *dst = dst0;
      |              ^~~~
/projects/github/ncs/zephyr/lib/os/memcpy32.c:21:20: error: initialization discards 'volatile' qualifier from pointer target type [-Werror=discarded-qualifiers]
   21 |  const char *src = src0;
      |                    ^~~~
cc1: all warnings being treated as errors
make[2]: *** [zephyr/CMakeFiles/zephyr.dir/build.make:398: zephyr/CMakeFiles/zephyr.dir/lib/os/memcpy32.c.obj] Error 1
make[1]: *** [CMakeFiles/Makefile2:2627: zephyr/CMakeFiles/zephyr.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```
